waitable_atomic/waitable_atomic.h

```diff
@@ -0,0 +1,324 @@
+#include <condition_variable>
+#include <mutex>
+
+namespace bricks {
+
+class CustomWaitableAtomicDestructor {
+ public:
+  virtual void WaitableAtomicDestructing() = 0;
+  virtual ~CustomWaitableAtomicDestructor() = default;
+};
+
+class ScopedUniqueLock {
+ public:
+  inline explicit ScopedUniqueLock(std::mutex& mutex) : lock_(mutex) {}
+  inline ~ScopedUniqueLock() {}
+  ScopedUniqueLock(ScopedUniqueLock&& rhs) { lock_ = std::move(rhs.lock_); }
+
+ private:
+  std::unique_lock<std::mutex> lock_;
+
+  ScopedUniqueLock(const ScopedUniqueLock&) = delete;
+  void operator=(const ScopedUniqueLock&) = delete;
+};
+
+class IntrusiveClient {
+ public:
+  class Interface {
+   public:
+    virtual ~Interface() {}
+
+   private:
+    virtual bool RefCounterTryIncrease() = 0;
+    virtual void RefCounterDecrease() = 0;
+    virtual bool IsNotDestructing() const = 0;
+    friend class IntrusiveClient;
+  };
+
+  inline explicit IntrusiveClient(Interface* object) : intrusive_object_(object) {
+    if (intrusive_object_) {
+      if (!intrusive_object_->RefCounterTryIncrease()) {
+        intrusive_object_ = nullptr;
+      }
+    }
+  }
+
+  IntrusiveClient(IntrusiveClient&& rhs) : intrusive_object_(nullptr) {
+    std::swap(intrusive_object_, rhs.intrusive_object_);
+  }
+
+  inline ~IntrusiveClient() {
+    if (intrusive_object_) {
+      intrusive_object_->RefCounterDecrease();
+    }
+  }
+
+  inline operator bool() const { return intrusive_object_ && intrusive_object_->IsNotDestructing(); }
+
+ private:
+  IntrusiveClient(const IntrusiveClient&) = delete;
+  void operator=(const IntrusiveClient&) = delete;
+
+  Interface* intrusive_object_;
+};
+
+template <typename DATA, bool INTRUSIVE = false>
+class WaitableAtomicImpl {
+ public:
+  class BasicImpl {
+   public:
+    typedef DATA T_DATA;
+    enum { IS_INTRUSIVE = false };
+
+    inline BasicImpl() : data_() {}
+
+    inline explicit BasicImpl(const T_DATA& data) : data_(data) {}
+
+    template <typename POINTER>
+    struct NotifyIfMutable {
+      class ImmutableAccessorDoesNotNotify {
+       public:
+        explicit ImmutableAccessorDoesNotNotify(POINTER*) {}
+      };
+      class MutableAccessorDoesNotify {
+       public:
+        explicit MutableAccessorDoesNotify(POINTER* parent) : parent_(parent), mark_as_unmodified_(false) {}
+        ~MutableAccessorDoesNotify() {
+          if (!mark_as_unmodified_) {
+            parent_->Notify();
+          }
+        }
+        void MarkAsUnmodified() { mark_as_unmodified_ = true; }
+
+       private:
+        POINTER* parent_;
+        bool mark_as_unmodified_;
+      };
+      typedef typename std::conditional<std::is_const<POINTER>::value,
+                                        ImmutableAccessorDoesNotNotify,
+                                        MutableAccessorDoesNotify>::type type;
+    };
+
+    template <class PARENT>
+    class TemplatedScopedAccessor : private ScopedUniqueLock, public NotifyIfMutable<PARENT>::type {
```

Comment:
the class (and especially all the places where it's used) is hard to follow, since it directly access some parents fields. if it accepted not only ```parent```, but ```mutex``` and ```pdata``` as constructor arguments, it'd be easier to read